### PR TITLE
Add-Entries Processor

### PR DESCRIPTION
**Pull request by shaavanga**
### Description
_Describe what this change achieves._
Added Add-Entry Processor Configuration
### Issues Resolved
_List any issues this PR will resolve, e.g. Closes [...]._
Add-Entry Processor Configuration

### Checklist
- [ ] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 1]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
 | :--- | :--- | :--- |
 | `entries` | Yes | A list of entries to add to an event. |
 | `key` | Yes | The key of the new entry to be added. Some examples of keys include `my_key`, `myKey`, and `object/sub_Key`. |
+| `metadata_key` | Yes | The key of the new metadata attribute to be added. Argument must be a literal string key and not JsonPointer. One of key or metadata_key is required. |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:21]:
```suggestion
| `metadata_key` | Yes | The key for the new metadata attribute. Argument must be a literal string key and not a `JsonPointer`. Either one string key or `metadata_key` is required. |
```

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 2]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
...
+| `format` | No | A format string to use as value of the new entry to be added. For example, ${key1}-${ke2} where key1 and key2 are existing keys in the event. Required if value is not specified. |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:22]: What do you mean by, "Required if value is not specified?"

```suggestion
| `format` | No | A format string to use as the value of the new entry, for example, `${key1}-${key2}` where `key1` and `key2` are existing keys in the event. Required if a value is not specified. |
```

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 3]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
...
+| `value_expression` | No | An expression string to use as value of the new entry to be added. For example, /key where key is an existing key in the event of type Number/String/Boolean. Expressions can also contain functions returning Number/String/Integer. For example length(/key) would return the length of the key in the event and key must of String type. Please see expressions syntax document for complete list of supported functions. Required if value and `format are not specified. |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:23]:
```suggestion
| `value_expression` | No | An expression string to use as the value of the new entry, for example, `/key` is an existing key in the event of type `Number/String/Boolean`. Expressions can also contain functions returning `Number/String/Integer`. For example, `length(/key)` would return the length of the key in the event when the key is a string. For more information on keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/). Required if value and `format are not specified. |
```

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 4]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
...
+| `add_when` | No | A Data Prepper ['/some_key == "test"'](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/) that will be evaluated to determine if the processor will be run on an Event. |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:24]: Is there a noun missing after "A Data Prepper..."?

**Review by Naarcha-AWS — CHANGES_REQUESTED** [review]

File: _data-prepper/pipelines/configuration/processors/add-entries.md (hunk 4 above)
Comment (shaavanga) [reply _data-prepper/pipelines/configuration/processors/add-entries.md:24]: No, Data prepper follows with conditional expression. It's right.

File: _data-prepper/pipelines/configuration/processors/add-entries.md (hunk 2 above)
Comment (shaavanga) [reply _data-prepper/pipelines/configuration/processors/add-entries.md]: `format` Required if `value` and `value_expression` is not specified

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 5]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
...
+| `metadata_key` | Yes | The key for the new metadata attribute. Argument must be a literal string key and not a `JsonPointer`. Either one string key or `metadata_key` is required. |
+| `format` | No | A format string to use as the value of the new entry, for example, `${key1}-${key2}` where `key1` and `key2` are existing keys in the event. Required if `value` and `value_expression` is not specified. |
+| `value_expression` | No | An expression string to use as the value of the new entry, for example, `/key` is an existing key in the event of type `Number/String/Boolean`. Expressions can also contain functions returning `Number/String/Integer`. For example, `length(/key)` would return the length of the key in the event when the key is a string. For more information on keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/). Required if value and `format are not specified. |
+| `add_when` | No | A Data Prepper ['/some_key == "test"'](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/) that will be evaluated to determine if the processor will be run on an Event. |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:24]:
```suggestion
| `add_when` | No | A [conditional expression](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/), such as [`/some_key == "test"'`](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/), that will be evaluated to determine if the processor will be run on the event. |
```

Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/processors/add-entries.md:24]: @shaavanga: Please confirm that the above change retains the technical meaning.

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 6]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
 | :--- | :--- | :--- |
 | `entries` | Yes | A list of entries to add to an event. |
 | `key` | Yes | The key of the new entry to be added. Some examples of keys include `my_key`, `myKey`, and `object/sub_Key`. |
+| `metadata_key` | Yes | The key for the new metadata attribute. Argument must be a literal string key and not a `JsonPointer`. Either one string key or `metadata_key` is required. |
+| `format` | No | A format string to use as the value of the new entry, for example, `${key1}-${key2}` where `key1` and `key2` are existing keys in the event. Required if `value` and `value_expression` is not specified. |
+| `value_expression` | No | An expression string to use as the value of the new entry, for example, `/key` is an existing key in the event of type `Number/String/Boolean`. Expressions can also contain functions returning `Number/String/Integer`. For example, `length(/key)` would return the length of the key in the event when the key is a string. For more information on keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/). Required if value and `format are not specified. |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:23]:
```suggestion
| `value_expression` | No | An expression string to use as the value of the new entry, for example, `/key` is an existing key in the event of type `Number/String/Boolean`. Expressions can also contain functions returning `Number/String/Integer`. For example, `length(/key)` would return the length of the key in the event when the key is a string. For more information about keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/). Required if value and `format are not specified. |
```

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 7]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
 | :--- | :--- | :--- |
 | `entries` | Yes | A list of entries to add to an event. |
 | `key` | Yes | The key of the new entry to be added. Some examples of keys include `my_key`, `myKey`, and `object/sub_Key`. |
+| `metadata_key` | Yes | The key for the new metadata attribute. Argument must be a literal string key and not a `JsonPointer`. Either one string key or `metadata_key` is required. |
+| `format` | No | A format string to use as the value of the new entry, for example, `${key1}-${key2}` where `key1` and `key2` are existing keys in the event. Required if `value` and `value_expression` is not specified. |
+| `value_expression` | No | An expression string to use as the value of the new entry, for example, `/key` is an existing key in the event of type `Number/String/Boolean`. Expressions can also contain functions returning `Number/String/Integer`. For example, `length(/key)` would return the length of the key in the event when the key is a string. For more information about keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/). Required if `value` and `format` are not specified. |
```
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:23]: I think we can remove the backticks around "Number/String/Boolean".

Perhaps reword to:

> an existing key in the event of with a type of either number, string, or boolean.

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 8]
```diff
@@ -36,6 +40,9 @@ pipeline:
         - key: "newMessage"
           value: 3
           overwrite_if_key_exists: true
+        - metadata_key: "myMetadataKey"
```
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:43]: This value doesn't require any quotes. Let's simplify it:

```
metadata_key: myMetadataKey
```

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 9]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
 | :--- | :--- | :--- |
 | `entries` | Yes | A list of entries to add to an event. |
 | `key` | Yes | The key of the new entry to be added. Some examples of keys include `my_key`, `myKey`, and `object/sub_Key`. |
+| `metadata_key` | Yes | The key for the new metadata attribute. Argument must be a literal string key and not a `JsonPointer`. Either one string key or `metadata_key` is required. |
```
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:21]: Let's remove the backticks from around "JsonPointer."

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: _data-prepper/pipelines/configuration/processors/add-entries.md (hunk 7 above)
Comment (shaavanga) [reply _data-prepper/pipelines/configuration/processors/add-entries.md]: Changed as suggested

File: _data-prepper/pipelines/configuration/processors/add-entries.md (hunk 8 above)
Comment (shaavanga) [reply _data-prepper/pipelines/configuration/processors/add-entries.md]: Changed as suggested

File: _data-prepper/pipelines/configuration/processors/add-entries.md (hunk 9 above)
Comment (shaavanga) [reply _data-prepper/pipelines/configuration/processors/add-entries.md]: Changed as suggested

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 10]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
 | :--- | :--- | :--- |
 | `entries` | Yes | A list of entries to add to an event. |
 | `key` | Yes | The key of the new entry to be added. Some examples of keys include `my_key`, `myKey`, and `object/sub_Key`. |
+| `metadata_key` | Yes | The key for the new metadata attribute. Argument must be a literal string key and not a JsonPointer. Either one string key or `metadata_key` is required. |
+| `format` | No | A format string to use as the value of the new entry, for example, `${key1}-${key2}` where `key1` and `key2` are existing keys in the event. Required if `value` and `value_expression` is not specified. |
+| `value_expression` | No | An expression string to use as the value of the new entry, for example, `/key` is an existing key in the event with a type of either number, string, or boolean. Expressions can also contain functions returning Number/String/Integer. For example, `length(/key)` would return the length of the key in the event when the key is a string. For more information about keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/). Required if `value` and `format` are not specified. |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:23]:
```suggestion
| `value_expression` | No | An expression string to use as the value of the new entry, for example, `/key` is an existing key in the event with a type of either a number, a string, or a Boolean. Expressions can also contain functions returning Number/String/Integer. For example, `length(/key)` would return the length of the key in the event when the key is a string. For more information about keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/). Required if `value` and `format` are not specified. |
```

**Review by Naarcha-AWS — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 11]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
 | :--- | :--- | :--- |
 | `entries` | Yes | A list of entries to add to an event. |
 | `key` | Yes | The key of the new entry to be added. Some examples of keys include `my_key`, `myKey`, and `object/sub_Key`. |
+| `metadata_key` | Yes | The key for the new metadata attribute. Argument must be a literal string key and not a JsonPointer. Either one string key or `metadata_key` is required. |
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:21]:
```suggestion
| `metadata_key` | Yes | The key for the new metadata attribute. The argument must be a literal string key and not a JSON Pointer. Either one string key or `metadata_key` is required. |
```

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 12]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
...
+| `format` | No | A format string to use as the value of the new entry, for example, `${key1}-${key2}` where `key1` and `key2` are existing keys in the event. Required if `value` and `value_expression` is not specified. |
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:22]:
```suggestion
| `format` | No | A format string to use as the value of the new entry, for example, `${key1}-${key2}`, where `key1` and `key2` are existing keys in the event. Required if `value` and `value_expression` is not specified. |
```

Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:22]: The last sentence needs some clarification. Do we mean "Required if neither `value` nor `value_expression` is specified" or "Required if both `value` and `value_expression` are not specified"?

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 13]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
...
+| `value_expression` | No | An expression string to use as the value of the new entry, for example, `/key` is an existing key in the event with a type of either a number, a string, or a Boolean. Expressions can also contain functions returning Number/String/Integer. For example, `length(/key)` would return the length of the key in the event when the key is a string. For more information about keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/). Required if `value` and `format` are not specified. |
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:23]:
```suggestion
| `value_expression` | No | An expression string to use as the value of the new entry. For example, `/key` is an existing key in the event with a type of either a number, a string, or a Boolean. Expressions can also contain functions returning Number/String/Integer. For example, `length(/key)` will return the length of the key in the event when the key is a string. For more information about keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/). Required if `value` and `format` are not specified. |
```

Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:23]: Same comment: "Required if neither `value` nor `format` is specified" or "Required if both `value` and `format` are not specified"? Also, is "Number/String/Integer" intentionally capitalized, or should it be lowercase?

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 14]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
...
+| `add_when` | No | A [conditional expression](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/), such as `/some_key == "test"'`, that will be evaluated to determine if the processor will be run on the event. |
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:24]:
```suggestion
| `add_when` | No | A [conditional expression](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/), such as `/some_key == "test"'`, that will be evaluated to determine whether the processor will be run on the event. |
```

**Review by natebower — COMMENTED** [review]
@shaavanga @Naarcha-AWS Please see my comments and changes and let me know if you have any questions. Thanks!

File: _data-prepper/pipelines/configuration/processors/add-entries.md (hunk 1 above)
Comment (shaavanga) [reply _data-prepper/pipelines/configuration/processors/add-entries.md]: Backticks are not required `JsonPointer`

File: _data-prepper/pipelines/configuration/processors/add-entries.md (hunk 2 above)
Comment (shaavanga) [reply _data-prepper/pipelines/configuration/processors/add-entries.md]: Backticks are not required for key1 and key2. Required if a value is not specified means `format` option is required If `value` and `value_expression` are not specified.

File: _data-prepper/pipelines/configuration/processors/add-entries.md (hunk 12 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/processors/add-entries.md:22]: The first one.

File: _data-prepper/pipelines/configuration/processors/add-entries.md [hunk 15]
```diff
@@ -18,6 +18,10 @@ You can configure the `add_entries` processor with the following options.
...
+| `metadata_key` | Yes | The key for the new metadata attribute. The argument must be a literal string key and not a JSON Pointer. Either one string key or `metadata_key` is required. |
+| `format` | No | A format string to use as the value of the new entry, for example, `${key1}-${key2}`, where `key1` and `key2` are existing keys in the event. Required if neither `value` nor `value_expression` is specified. |
+| `value_expression` | No | An expression string to use as the value of the new entry. For example, `/key` is an existing key in the event with a type of either a number, a string, or a Boolean. Expressions can also contain functions returning number/string/integer. For example, `length(/key)` will return the length of the key in the event when the key is a string. For more information about keys, see [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/).  |
+| `add_when` | No | A [conditional expression](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/), such as `/some_key == "test"'`, that will be evaluated to determine whether the processor will be run on the event. |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/add-entries.md:24]:
```suggestion
| `add_when` | No | A [conditional expression](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/), such as `/some-key == "test"'`, that will be evaluated to determine whether the processor will be run on the event. |
```

**Pull request merged**